### PR TITLE
Merge `BuildWorldChildren` and `BuildChildren` traits.

### PR DESCRIPTION
**Pull request by yrns**
# Objective

The `BuildChildren` and `BuildWorldChildren` traits are mostly identical, so I decided to try and merge them. I'm not sure of the history, maybe they were added before GATs existed.

## Solution

- Add an associated type to `BuildChildren` which reflects the prior differences between the `BuildChildren` and `BuildWorldChildren` traits.
- Add `ChildBuild` trait that is the bounds for `BuildChildren::Builder`, with impls for `ChildBuilder` and `WorldChildBuilder`.
- Remove `BuildWorldChildren` trait and replace it with an impl of `BuildChildren` for `EntityWorldMut`.

## Testing

I ran several of the examples that use entity hierarchies, mainly UI.

---

## Changelog

n/a

## Migration Guide

n/a

### CODE REVIEW

File: crates/bevy_hierarchy/src/child_builder.rs [hunk 1]
```diff
@@ -279,39 +279,62 @@ pub struct ChildBuilder<'a> {
     push_children: PushChildren,
 }
 
-impl ChildBuilder<'_> {
+/// Trait for building children entities and adding them to a parent entity. See
+/// [`BuildChildren::with_children`].
+pub trait ChildBuild {
+    /// Spawn output type.
```
Comment (alice-i-cecile) [inline crates/bevy_hierarchy/src/child_builder.rs:285]: This doc comment needs to be clearer.

File: crates/bevy_hierarchy/src/child_builder.rs [hunk 2]
```diff
@@ -279,39 +279,62 @@ pub struct ChildBuilder<'a> {
     push_children: PushChildren,
 }
 
-impl ChildBuilder<'_> {
+/// Trait for building children entities and adding them to a parent entity. See
```
Comment (alice-i-cecile) [inline crates/bevy_hierarchy/src/child_builder.rs:282]: The "see" comment is unclear to readers. You should explain how these are related.

File: crates/bevy_hierarchy/src/child_builder.rs (hunk 1 above)
Comment (alice-i-cecile) [inline crates/bevy_hierarchy/src/child_builder.rs:285]: A doc comment on the meaning of the lifetime where clause would really help clarity.

**Comment by alice-i-cecile** [comment]
Lovely, I was hoping we could do this.

**Review by alice-i-cecile — APPROVED** [review]
I like the changes. Thanks!

File: crates/bevy_hierarchy/src/child_builder.rs [hunk 3]
```diff
@@ -279,39 +279,68 @@ pub struct ChildBuilder<'a> {
...
-impl ChildBuilder<'_> {
+/// Trait for building children entities and adding them to a parent entity. This is used in
+/// implementations of [`BuildChildren`] as a bound on the [`Builder`](BuildChildren::Builder)
+/// associated type. The closure passed to [`BuildChildren::with_children`] accepts an
+/// implementation of `ChildBuild` so that children can be spawned via [`ChildBuild::spawn`].
+pub trait ChildBuild {
+    /// Spawn output type. Both [`spawn`](Self::spawn) and [`spawn_empty`](Self::spawn_empty) return
+    /// an implementation of this type so that children can be operated on via method-chaining.
+    /// Implementations of `ChildBuild` reborrow `self` when spawning entities (see
+    /// [`Commands::spawn_empty`] and [`World::get_entity_mut`]). Lifetime `'a` corresponds this
```
Comment (janhohenheim) [inline crates/bevy_hierarchy/src/child_builder.rs:290]:
```suggestion
    /// [`Commands::spawn_empty`] and [`World::get_entity_mut`]). Lifetime `'a` corresponds to this
```

File: crates/bevy_hierarchy/src/child_builder.rs [hunk 4]
```diff
@@ -279,39 +279,68 @@ pub struct ChildBuilder<'a> {
     push_children: PushChildren,
 }
 
-impl ChildBuilder<'_> {
+/// Trait for building children entities and adding them to a parent entity. This is used in
+/// implementations of [`BuildChildren`] as a bound on the [`Builder`](BuildChildren::Builder)
+/// associated type. The closure passed to [`BuildChildren::with_children`] accepts an
+/// implementation of `ChildBuild` so that children can be spawned via [`ChildBuild::spawn`].
+pub trait ChildBuild {
```
Comment (janhohenheim) [inline crates/bevy_hierarchy/src/child_builder.rs:286]: Worth thinking about whether we want the user to be able to implement this trait on their own. The current API implies this and makes adding a new method a breaking change.
What do you think @alice-i-cecile?

**Review by janhohenheim — CHANGES_REQUESTED** [review]
Great idea, thanks! Just some minor notes.

File: crates/bevy_hierarchy/src/child_builder.rs (hunk 3 above)
Comment (yrns) [reply crates/bevy_hierarchy/src/child_builder.rs]: Thanks.

File: crates/bevy_hierarchy/src/child_builder.rs (hunk 4 above)
Comment (alice-i-cecile) [reply crates/bevy_hierarchy/src/child_builder.rs:286]: IMO we don't bother sealing this but we also just don't really care about adding new methods. This isn't long for the world: complexity isn't worth it here.

**Review by janhohenheim — APPROVED** [review]

**Review by JoJoJet — APPROVED** [review]

**Pull request merged**